### PR TITLE
[GLUTEN-5352][GLUTEN-5459][CH]Fix and improve year function

### PR DESCRIPTION
**Pull request by KevinyhZou**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#5352, #5459)

## How was this patch tested?
TEST BY UT

性能测试：

### 端到端SQL测试
表test_tbl(d string), 数据量 3000W行，测试SQL `select count(1) from test_tbl where year(d) > '1990'`
PR改动前：5.245s， 5.442s， 5.458s；
PR改动后：4.689s，4.782s， 5.011s;

### CODE REVIEW

File: cpp-ch/local-engine/Functions/FunctionGetDateData.h [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+namespace local_engine
+{
+template<bool get_year, bool get_date, typename T>
+class FunctionGetDateData : public DB::IFunction
+{
+public:
+    FunctionGetDateData() = default;
+    ~FunctionGetDateData() override = default;
+    bool checkAndGetDateData(DB::ReadBuffer & buf, size_t buf_size, T &x, const DateLUTImpl & date_lut, UInt8 & can_be_parsed) const
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Functions/FunctionGetDateData.h:48]: 这个函数实现缺少内存越界检查，或许不应该设置为public

File: cpp-ch/local-engine/Functions/FunctionGetDateData.h [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    {
+        auto checkNumbericASCII = [&](DB::ReadBuffer & rb, size_t start, size_t length) -> bool
+        {
+            for (size_t i = start; i < start + length; ++i)
+            {
+                if (!isNumericASCII(*(rb.position() + i)))
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Functions/FunctionGetDateData.h:54]: 对readbuffer内部缓冲区的访问，需要检查是否越界

File: cpp-ch/local-engine/Functions/FunctionGetDateData.h [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                {
+                    return false;
+                }
+            }
+            return true;
+        };
+        auto checkDelimiter = [&](DB::ReadBuffer & rb, size_t pos) -> bool
+        {
+            if (*(rb.position() + pos) != '-')
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Functions/FunctionGetDateData.h:63]: 同上

File: cpp-ch/local-engine/Functions/FunctionGetDateData.h [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                while(!buf.eof() && *buf.position() == ' ')
+                {
+                    buf.position() ++;
+                }
+                if(buf.buffer().end() - buf.position() < 4)
+                {
+                    null_container[i] = true;
+                    continue;
+                }
+                UInt8 can_be_parsed = 1;
```
Comment (liuneng1994) [inline cpp-ch/local-engine/Functions/FunctionGetDateData.h:152]: use bool

**Comment by KevinyhZou** [comment]
done

**Review by liuneng1994 — APPROVED** [review]
LGTM

**Pull request merged**